expo/expo
# fix(metro-config/serializer): include moduleID wrapper when hashing chunks

# Why

Currently metro is hashing chunks of JS without taking into account moduleIDs. 
When just the moduleID of chunk changes, hash becomes obsolete but it doesn't change. That leads to caching bug, more deeply explained in: https://github.com/expo/expo/issues/30203

ex: we have dynamically imported module: `index-8937624c6fb773d1d9cacd4db3808a3d.js` with moduleID: `3679`. Some other part of the app changes, and moduleID changes to "3680". Hash doesn't change when in fact content of the file changed with new references to moduleIDs.

```
  ~/Downloads diff index-8937624c6fb773d1d9cacd4db3808a3d.js index-8937624c6fb773d1d9cacd4db3808a3d-new.js
471c471
< ), 3679, [3680, 14]);
---
> ), 3676, [3677, 14]);
12476c12476
< ), 3680, []);
---
> ), 3677, []);
➜  ~/Downloads
```

# How

Modified `getStableChunkSource` function in `@expo/metro-config/serializer/serializeChunks.js` file so it takes `moduleIDs` wrapping into account.

# Test Plan

Regression testing:
1. Run `npx expo export -p web -c` in `bare-expo` a few times. Hashes should always be the same without changes.
2. Add some console.log(). Hash should change.

Feature testing: 
1. Download prepared testing environment https://github.com/Nsquik/expo/tree/nsquik/serialize-chunks-with-wrapping-testing.
2. Open `bare-expo` app.
3. Run to generate chunks with indexes: `npx expo export -p web -c`. You should pay attention to dist/_expo/static/js/web/LazyLoadedComponent-e7dc611b96c4f39cfe473a9621c9a1d0.js. It gets assigned moduleID: "2627"
4. Now add new module to application. Import `NotLazyLoadedComponent` to MainNavigator like this: 
![Screenshot 2024-07-05 at 16 46 45](https://github.com/expo/expo/assets/46610139/14ebce67-b4d8-4516-900f-74beaa498c5f)
5. Now moduleID of LazyLoadedComponent increments by 1. There is one more module imported to app before LazyLoadedComponent.
6. Run to generate chunks with indexes: `npx expo export -p web -c`. Hash should change to: LazyLoadedComponent-c842e90a6b13b22bafa74fc6a9d123fe.js + moduleID: 2628.

# Checklist
